### PR TITLE
Replace MethodIDs with ImageIDs 

### PR DESCRIPTION
**Pull request by flaub**

### CODE REVIEW

**Review by tzerrell — APPROVED** [review]
🎉

LGTM (of course, I'm not really suited for reviewing the more circuit-integration-y parts of this, but the bits I could review well look good)

File: risc0/build/risc0.ld [hunk 1]
```diff
@@ -21,10 +21,9 @@ EXTERN(__start)
...
 /* Must match risc0/zkvm/platform/src/memory.rs */
 MEMORY {
-  stack        : ORIGIN = 0x00000000, LENGTH =  16M
-  data    (RW) : ORIGIN = 0x01000000, LENGTH =  16M
-  heap         : ORIGIN = 0x02000000, LENGTH =  32M
-  input        : ORIGIN = 0x04000000, LENGTH =  32M
+  stack        : ORIGIN = 0x02000000, LENGTH =  16M
+  data    (RW) : ORIGIN = 0x03000000, LENGTH =  16M
+  heap         : ORIGIN = 0x04000000, LENGTH =  32M
   prog    (X)  : ORIGIN = 0x06000000, LENGTH =  32M
```
Comment (jbruestle) [inline risc0/build/risc0.ld:27]: Is there a reason not to put prog @ 0x08... - 0x0c...?  I was of the impression that system memory starts only after 0x0c and that nondet was now below @0x02.  It seems like 64M for heap and 64M for code would be better than 32M for heap and 96M for program.  Or maybe even more room for heap.

File: risc0/zkp/src/verify/mod.rs [hunk 2]
```diff
@@ -258,13 +250,12 @@ pub fn verify<'a, H, C, F>(
     hal: &'a H,
     circuit: &C,
     seal: &'a [u32],
-    journal: &'a [u32],
-    check_code: F,
+    check_globals: F,
```
Comment (jbruestle) [inline risc0/zkp/src/verify/mod.rs:253]: We really need to support both checking globals and checking the code root.  I'm OK with the ZKVM temporarily ignoring the code root until we can add the build support to precompute the valid roots, but the recursion support will also need to check the code root, so I'd like to keep the API in place.

File: risc0/zkvm/platform/src/memory.rs [hunk 3]
```diff
@@ -54,17 +54,14 @@ impl Region {
...
-pub const DATA: Region = Region::new(0x0100_0000, mb(16));
-pub const HEAP: Region = Region::new(0x0200_0000, mb(32));
-pub const INPUT: Region = Region::new(0x0400_0000, mb(32));
-pub const PROG: Region = Region::new(0x0600_0000, mb(32));
+pub const INPUT: Region = Region::new(0x0000_0400, mb(32) - kb(1));
+pub const STACK: Region = Region::new(0x0200_0000, mb(16));
+pub const DATA: Region = Region::new(0x0300_0000, mb(16));
+pub const HEAP: Region = Region::new(0x0400_0000, mb(32));
+pub const TEXT: Region = Region::new(0x0600_0000, mb(32));
 pub const OUTPUT: Region = Region::new(0x0800_0000, mb(32));
```
Comment (jbruestle) [inline risc0/zkvm/platform/src/memory.rs:62]: Is there any reason to have a specific output + commit region anymore?  As far as I can tell there is no reason output can't just send heap pointers to the host.  Also for the commits, cant we just have a single SHA256 state and a single 16 word staging buffer (both in the heap), and just write words to the staging buffer and kick off a compress whenever it's full?

Comment (flaub) [reply risc0/zkvm/platform/src/memory.rs:62]: Ah yes, that's a good point on both counts. Now that we have `sys_sha_buffer`, we can probably drop the COMMIT region. I also don't think we need an OUTPUT region anymore!

File: risc0/build/risc0.ld (hunk 1 above)
Comment (flaub) [reply risc0/build/risc0.ld:27]: That all makes great sense. I was also thinking maybe `prog` should live above the system region? Like maybe reading from the PC could use a special memory op.

**Review by jbruestle — APPROVED** [review]

**Review by SchmErik — APPROVED** [review]

File: risc0/build/src/lib.rs [hunk 4]
```diff
@@ -121,57 +121,21 @@ impl Risc0Method {
             std::process::exit(-1);
         }
 
-        let method_id_path = self.elf_path.with_extension("id");
-        let elf_sha_path = self.elf_path.with_extension("sha");
-        let elf_contents = std::fs::read(&self.elf_path).unwrap();
-
-        let elf_sha = Sha256::new()
-            // Method ID calculation is slow, so only recalculate it if we
-            // actually get a different ELF file.
```
[the remark this answers is not in the document]
Comment (flaub) [reply risc0/build/src/lib.rs:130]: It seems to be pretty fast without having to cache it. Caching comes with all sorts of complexity and potential for bugs. If we find out it's still too slow we can bring caching back.

**Pull request merged**